tensorflow/core/kernels/segment_reduction_ops.cc

```diff
@@ -288,27 +313,66 @@ class UnsortedSegmentSumOp : public OpKernel {
     auto output_flat = output->flat_outer_dims<T>();
 
     auto data_ptr = data.template flat<T>().data();
-    functor::UnsortedSegmentSumFunctor<Device, T, Index>()(
-        context, context->template eigen_device<Device>(), output_rows,
+    ReductionFunction(context, context->template eigen_device<Device>(), output_rows,
         segment_ids.shape(), segment_flat, data.NumElements(), data_ptr,
         output_flat);
   }
+private:
+  Functype ReductionFunction;
 };
 
-#define REGISTER_CPU_UNSORTED_KERNELS(type, index_type)                \
+template <typename Device, class T, class Index>
+class UnsortedSegmentSumOp
+  : public UnsortedSegmentBaseOp<Device, T, Index,
+                                   functor::UnsortedSegmentSumFunctor<Device, T, Index>> {
+public:
+  explicit UnsortedSegmentSumOp(OpKernelConstruction* context)
+    : UnsortedSegmentBaseOp<Device, T, Index, functor::UnsortedSegmentSumFunctor<Device, T, Index>>(context, functor::UnsortedSegmentSumFunctor<Device, T, Index>()) {}
+};
+
+template <typename Device, class T, class Index>
+class UnsortedSegmentMaxOp
+  : public UnsortedSegmentBaseOp<Device, T, Index,
+                functor::UnsortedSegmentMaxFunctor<Device, T, Index>> {
```

Comment:
Can we just pass the functor into the base class constructor here (i.e. from the derived classes constructor), rather than passing it as a template argument? It looks like the constructor for UnsortedSegmentBaseOp takes the functor as an arg.